cloud/services/subnets/subnets_test.go

```diff
@@ -0,0 +1,243 @@
+*/
+
+package subnets
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/golang/mock/gomock"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha2"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/routetables/mock_routetables"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/securitygroups/mock_securitygroups"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/subnets/mock_subnets"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestReconcileSubnets(t *testing.T) {
+	testcases := []struct {
+		name        string
+		subnetSpec  Spec
+		vnetSpec    *infrav1.VnetSpec
+		expectError bool
+		expect      func(m *mock_subnets.MockClientMockRecorder, m1 *mock_routetables.MockClientMockRecorder, m2 *mock_securitygroups.MockClientMockRecorder)
+	}{
+		{
+			name: "subnet does not exist",
+			subnetSpec: Spec{
+				Name:                "my-subnet",
+				CIDR:                "10.0.0.0/16",
+				VnetName:            "my-vnet",
+				RouteTableName:      "my-subent_route_table",
+				SecurityGroupName:   "my-sg",
+				Role:                infrav1.SubnetNode,
+				InternalLBIPAddress: "10.0.0.10",
+			},
+			vnetSpec:    &infrav1.VnetSpec{Name: "my-vnet"},
+			expectError: false,
+			expect: func(m *mock_subnets.MockClientMockRecorder, m1 *mock_routetables.MockClientMockRecorder, m2 *mock_securitygroups.MockClientMockRecorder) {
+				m.Get(context.TODO(), "", "my-vnet", "my-subnet").
+					Return(network.Subnet{}, autorest.NewErrorWithResponse("", "", &http.Response{StatusCode: 404}, "Not found"))
+
+				m1.Get(context.TODO(), "my-rg", "my-subent_route_table").
+					Return(network.RouteTable{}, nil)
+
+				m2.Get(context.TODO(), "my-rg", "my-sg").
+					Return(network.SecurityGroup{}, nil)
+
+				m.CreateOrUpdate(context.TODO(), "", "my-vnet", "my-subnet", gomock.AssignableToTypeOf(network.Subnet{}))
+			},
+		},
+		{
+			name: "vnet was provided but subnet is missing",
+			subnetSpec: Spec{
+				Name:                "my-subnet",
+				CIDR:                "10.0.0.0/16",
+				VnetName:            "custom-vnet",
+				RouteTableName:      "my-subent_route_table",
+				SecurityGroupName:   "my-sg",
+				Role:                infrav1.SubnetNode,
+				InternalLBIPAddress: "10.0.0.10",
+			},
+			vnetSpec:    &infrav1.VnetSpec{ResourceGroup: "custom-vnet-rg", Name: "custom-vnet", ID: "id1"},
+			expectError: true,
+			expect: func(m *mock_subnets.MockClientMockRecorder, m1 *mock_routetables.MockClientMockRecorder, m2 *mock_securitygroups.MockClientMockRecorder) {
+				m.Get(context.TODO(), "custom-vnet-rg", "custom-vnet", "my-subnet").
+					Return(network.Subnet{}, errors.New("vnet was provided but subnet my-subnet is missing"))
```

Comment:
this error won't be return by Get(), it will be returned by Reconcile(). Get() will return a not found error. What's important is that expectedError above is true.

Reply:
so its okay to just check that an error happened, or should it also check for the exact error?

Reply:
I was thinking the same thing... maybe we should change expected error to a string

Reply:
how about we change it to an actual error and check that the error matches?

Reply:
even better!

Reply:
ok, will try that! Thanks